js/hard-mode.js

```diff
@@ -0,0 +1,116 @@
+
+    /**
+    * Gets all empty spaces on the board, then checks each winning combination.
+    * If two empty spaces are found, move on to the next combination.
+    * If one empty space is found and the other two spaces are occupied by different markers, move on.
+    * If no empty spaces are found, move on to the next combination.
+    * If one empty space is found and the other two spaces are occupied by the same marker, return the empty Id, since
+    * that is the winning move - either for the computer to win now, or to block the player from winning on their next
+    * turn.
+    */
+    HardMode.getWinningId = function() {
+        var emptyIds = Array.from(document.querySelectorAll('.' + EMPTY_SPACE_CLASS)).map(function(emptySpace) {
+            return emptySpace.id;
+        });
+        var markerById = {};
+        for (var i = 0; i < WINNING_COMBINATIONS.length; i++) {
+            var firstOccupiedSpaceMarker = null;
+            var possibleWinningId = null;
+            var winningCombination = WINNING_COMBINATIONS[i];
+            for (var j = 0; j < winningCombination.length; j++) {
+                var spaceId = winningCombination[j];
+                if (emptyIds.includes(spaceId)) {
+                    if (possibleWinningId !== null) {
+                        possibleWinningId = null;
+                        break;
+                    }
+                    possibleWinningId = spaceId;
+                } else {
+                    var occupyingMarker = markerById[spaceId];
+                    if (occupyingMarker === undefined) {
+                        occupyingMarker = document.getElementById(spaceId).className;
+                        markerById[spaceId] = occupyingMarker;
+                    }
+                    if (firstOccupiedSpaceMarker === null) {
+                        firstOccupiedSpaceMarker = occupyingMarker;
+                    } else if (firstOccupiedSpaceMarker !== occupyingMarker) {
+                        possibleWinningId = null;
+                        break;
+                    }
+                }
+            }
+            if (possibleWinningId !== null) {
+                return possibleWinningId;
+            }
+        }
+    };
+
+    /**
+    * Determines an element for the computer to place a marker in, according to the following priority:
+    * 1. Winning move (either for the computer or to block the player)
+    * 2. Center space
+    * 3. Corner space
+    * 4. Any remaining available space
+    */
+    HardMode.computerMove = function() {
+        var winningId = HardMode.getWinningId();
+        if (winningId !== undefined) {
+            return document.getElementById(winningId);
+        }
+
+        var centerSpace = document.getElementById('e-box');
+        if (centerSpace.className === EMPTY_SPACE_CLASS) {
+            return centerSpace;
+        }
+
+        var emptyCornerSelectors = ['a-box', 'c-box', 'g-box', 'i-box'].map(function(id) {

```

Comment:
The `-box` part of each of these elements is repetitive and can be moved into the function passed into .map.